### PR TITLE
[NewOptimizer] Port #26826 to new optimizer

### PR DESCRIPTION
**Pull request by jrevels**

### CODE REVIEW

File: base/compiler/ssair/inlining2.jl
Comment (jrevels) [reply base/compiler/ssair/inlining2.jl]: Oh, I forgot to check `def.n`. Let me do that...

File: base/compiler/ssair/inlining2.jl
Comment (Keno) [inline base/compiler/ssair/inlining2.jl]: If this is what's used for the cache, is the below atypes actually used anywhere?

**Comment by jrevels** [comment]
> Seems like f(args...) = g(args..., 1) could benefit from this as well.

Ah, good point.

File: base/compiler/ssair/inlining2.jl
Comment (jrevels) [reply base/compiler/ssair/inlining2.jl]: Oh neat, you're right - the modified `atypes` was used elsewhere in the old inlining code, but not here. I guess I can just remove that code then.

**Review by Keno — APPROVED** [review]
LGTM if tests pass with this and the new optimizer enabled.

File: base/compiler/ssair/inlining2.jl
Comment (Keno) [inline base/compiler/ssair/inlining2.jl]: `isva` is used below still.

**Review by Keno — CHANGES_REQUESTED** [review]
A bunch of thing seems to go of the rails when I run the tests with this PR.

**Comment by jrevels** [comment]
> A bunch of thing seems to go of the rails when I run the tests with this PR.

Just started a local build + tests with the new optimizer enabled, anything I should watch out for? Do all tests on master pass with the new optimizer enabled (besides the ones from #26826, which should require this port)?

**Comment by Keno** [comment]
Yes, tests pass on master.

**Comment by jrevels** [comment]
Okay, I ran tests locally, and it seems there's only one thing that's failing: 

```
SparseArrays/higherorderfns: Test Failed at /Users/jarrettrevels/data/repos/julia-dev/usr/share/julia/stdlib/v0.7/SparseArrays/test/higherorderfns.jl:269
  Expression: #= /Users/jarrettrevels/data/repos/julia-dev/usr/share/julia/stdlib/v0.7/SparseArrays/test/higherorderfns.jl:269 =# @allocated(broadcast!(f, Q, X, Y, Z)) == 0
   Evaluated: 16 == 0
```

This is [the broken sparse arrays broadcast test that #26826 allowed us to mark as unbroken](https://github.com/JuliaLang/julia/commit/429a885e74e8327e6e257f3625aa5c90378b2483#diff-b3c560f5f36016969257fb59e505d16cL269). I guess it's still broken with the new optimizer. We could investigate why, and in the meantime just mark it `@test_broken` like it was before #26826.

**Comment by Keno** [comment]
Is that with this PR merged? That's what I saw before I merged this PR.

**Comment by jrevels** [comment]
That's from running `make test` on 7506dc1c3b89f272aaa7027f4ba30b999f2c760e with

```
diff --git a/base/compiler/optimize.jl b/base/compiler/optimize.jl
index 5943037..73fba2c 100644
--- a/base/compiler/optimize.jl
+++ b/base/compiler/optimize.jl
@@ -261,7 +261,7 @@ function isinlineable(m::Method, src::CodeInfo, mod::Module, params::Params, bon
     return inlineable
 end

-const enable_new_optimizer = RefValue(false)
+const enable_new_optimizer = RefValue(true)

 # converge the optimization work
 function optimize(me::InferenceState)
```

I could run it on the merge commit as well, did that change things (i.e. have changes to master since 429a885e74e8327e6e257f3625aa5c90378b2483 broken other tests w.r.t. the new optimizer when combined with this PR)?

A summary for anybody following along (cc @vchuravy, since you asked about the relevant test in #26989): 

#26826 added some compiler tests that should've failed under the new optimizer before this PR, and this PR allows those tests to pass w/ the new optimizer enabled. Furthermore, #26826 fixed the aforementioned `@test_broken` sparse broadcast test as a side effect, so #26826 changed it from `@test_broken` to `@test`. However, it seems that this test is still broken under the new optimizer, even with this PR.

**Comment by Keno** [comment]
Seems like I may have had other thing locally mess up my test results. Merging.

**Pull request merged**
